src/wp-includes/blocks.php

```diff
@@ -387,49 +404,60 @@ function get_block_metadata_i18n_schema() {
  * @since 6.5.0 Added support for `allowedBlocks`, `viewScriptModule`, and `viewStyle` fields.
  * @since 6.7.0 Allow PHP filename as `variations` argument.
  *
- * @param string $file_or_folder Path to the JSON file with metadata definition for
+ * @param string $file_or_metadata_source Path to the JSON file with metadata definition for
```

Comment:
Thanks, this is a good idea and I'll give it a shot. One thing I want to nail down is the matching of specific blocks from the registered collection and later `register_block_type_from_metadata()` calls.

To use a concrete example, we might call:
```php
WP_Block_Metadata_Registry::register_collection(
  '/srv/http/wordpress/wp-includes/blocks/',
  '/srv/http/wordpress/wp-includes/blocks/blocks-json.php'
)
```

The `blocks-json.php` file is unchanged and is formatted like:
```php
<?php return array(
  'archives' => array(
    'name' => 'core/archives',
    'title' => 'Archives',
	// ..properties omitted..
  ),
  'audio' => array(
    'name' => 'core/audio',
    'title' => 'Audio',
	// ..properties omitted..
  ),
  'avatar' => array(
    'name' => 'core/avatar',
    'title' => 'Avatar',
	// ..properties omitted..
  ),
  // ..other blocks..
);
```

In this case, to match up block names, we would look for calls like:
```php
register_block_type_from_metadata( '/srv/http/wordpress/wp-includes/blocks/archives' );
OR
register_block_type_from_metadata( '/srv/http/wordpress/wp-includes/blocks/archives.json' );
```

In each of these cases we would:
 * Extract the last directory of the path (`archives`) and use this as our key
 * Extract the path except the last directory (`/srv/http/wordpress/wp-includes/blocks`) and use this as a collection

For the collection, we'd check if the collection is registered and if the data read from the provided PHP file had an array key matching the last part of our path (`archives`), and whatever is found there would be used as the block metadata.

Does that sound right? I considered some other ideas, like matching block names and looping through collections looking for them, but that doesn't seem to work, especially with core and the Gutenberg plugin registering different metadata for blocks with the same name.

_( The only tricky bit I've found so far is that the Gutenberg plugin does some registrations with `../` in them, like `/srv/http/wordpress/wp-content/plugins/gutenberg/lib/../build/block-library/blocks/group/block.json`, while others do not, like `/srv/http/wordpress/wp-content/plugins/gutenberg/build/block-library/blocks/categories`. )_

Reply:
Great question, I hadn't yet thought about how to map the PHP file contents back to the specific block name.

I think the following would work well:
* If registered with something like `/srv/blocks/archives/block.json` or just a directory `/src/blocks/archives` (in which case the file name of `block.json` is assumed), we use the last directory (i.e. `archives` in this example).
* If registered with a path to a JSON file that is not `block.json`, we use the name of that file, e.g. `/srv/blocks/button.json` will lead to using `button`.
* These would work well as defaults, but I think we should also support alternatives since both of the above are, at the end of the day, still just assumptions how people will use this. So I think a good idea would be for `WP_Block_Metadata_Registry::register_collection()` to support an optional argument for a callback on how to parse a block JSON file name into the block's identifier within the built PHP file array. This would be more advanced to use, but the good thing is that most developers wouldn't need it since they probably use one of the two previous approaches that we would support by default.

For knowing which collection a specific block JSON file belongs to, I think we could always rely on a check whether it starts with a specific collection's registered root directory. We could add some safe guards to `WP_Block_Metadata_Registry::register_collection()` so that you can't register a path that will cause problems (for example _nobody_ should ever register `WP_PLUGIN_DIR` as a collection, as that would mean every single block in every plugin would fall under this hypothetical collection. In other words, only a path within a specific plugin directory should be allowed in this context (e.g. `WP_PLUGIN_DIR . '/gutenberg/blocks'`).

Reply:
I've just pushed an implementation of the 'last directory' idea to this PR. In my testing, it works for all of core, and about 80% of WooCommerce. The remaining 20% that doesn't work are in two tiered directories, I think.

Will look at some of these other ideas tomorrow.

Reply:
If I'm reading the current version of the function correctly, when passing a file or directory, it only supports `block.json` files or directories [[link](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks.php#L410-L412)]. It does not support files like `arbitrary_string.json`. Keeping this constraint might allow the new code to be simpler and avoid disk access (ie, not checking is_dir()).

Reply:
Ah you're right, good point. In that case, it's probably fine to always go with the last directory above the `block.json` file to get the block name as the default behavior. We could add support for an optional custom callback to manually determine this later, doesn't have to be in the first iteration of the API.